src/components/pages/media-gallery/create/uploader/media-gallery-upload-provider.tsx

```diff
@@ -0,0 +1,232 @@
+
+  const { t } = useTranslation();
+
+  const { add, getOneByKey, getManyByKey, deleteByID, update } =
+    useIndexedDBStore<ResourceMediaGallery>(STORE.ASSETS);
+
+  const reFetchAssets = async () => {
+    const allUnUsedAssets = await getManyByKey("isUsed", 0);
+    console.warn(" allUnUsedAssets", allUnUsedAssets);
+    setAssets((_draft) => {
+      _draft.a = allUnUsedAssets.sort((a, b) => b[resourcesSortBy] - a[resourcesSortBy]);
+    });
+  };
+
+  useEffect(() => {
+    reFetchAssets();
+  }, [resourcesSortBy]);
+
+  const fetchMyUploads = async () => {
+    const { data } = await axListMyUploads();
+
+    // housekeeping for expired assets
+    const newAssetsHashKeys = data.map((a) => a.hashKey);
+    const allUnUsedAssets = await getManyByKey("isUsed", 0);
+    for (const asset of allUnUsedAssets) {
+      if (!newAssetsHashKeys.includes(asset.hashKey) && asset.status === AssetStatus.Uploaded) {
+        await deleteByID(asset.id);
+      }
+    }
+
+    // Update all fetched assets into IndexedDB
+    for (const asset of data) {
+      const dbAsset = await getOneByKey("hashKey", asset.hashKey);
+      if (!dbAsset) {
+        await add({
+          ...asset,
+          status: AssetStatus.Uploaded,
+          url: null,
+          contributor: "",
+          caption: "",
+          rating: 0,
+          licenseId: SITE_CONFIG.LICENSE.DEFAULT,
+          isUsed: 0,
+        });
+      }
+    }
+
+    await reFetchAssets();
+  };
+
+  useEffect(() => {
+    setupDB();
+    fetchMyUploads();
+  }, []);
+
+  const updateLocalAssetStatus = async (hashKey, status: AssetStatus) => {
+    setAssets((_draft) => {
+      const index = _draft.a.findIndex((a) => a.hashKey === hashKey);
+      if (index > -1) {
+        _draft.a[index].status = status;
+      }
+    });
+    setMediaGalleryAssets((_draft) => {
+      const index = _draft.a.findIndex((a) => a.hashKey === hashKey);
+      if (index > -1) {
+        _draft.a[index].status = status;
+      }
+    });
+  };
+
+  const handleMediaFiles = async (pendingResource, noSave) => {
+    try {
+      const r = await axUploadMediaGalleryResource(pendingResource);
+      if (r.success && noSave) {
+        await update({
+          ...pendingResource,
+          status: AssetStatus.Uploaded,
+        });
+        await updateLocalAssetStatus(pendingResource.hashKey, AssetStatus.Uploaded);
+      }
+    } catch (e) {
+      console.error(e);
+      if (noSave) {
+        await updateLocalAssetStatus(pendingResource.hashKey, AssetStatus.Pending);
+      }
+    }
+  };
+
+  const uploadPendingResource = async (pendingResource, noSave = true) => {
+    if (noSave) {
+      await updateLocalAssetStatus(pendingResource.hashKey, AssetStatus.InProgress);
+    }
+
+    handleMediaFiles(pendingResource, noSave);
+  };
+
+  const tryResourceSync = async () => {
+    const pendingResources = await getManyByKey("status", AssetStatus.Pending);
+    for (const pendingResource of pendingResources) {
```

Comment:
can we use Promise.all instead of calling the endpoint one-by-one in a for loop?

Reply:
yes we can use Promise.all and it will improve the performance , updating the code here and also this part of code is taken from **use-observation-resources** in biodvi-ui , we can update there too if we want .